### PR TITLE
Av 221779 Fix secret and gateway re-ordering

### PR DESCRIPTION
**Pull request by akshayhavile**
This PR contains fix for 2 issues:
1. [AV-221779] : Fix secret and gateway re-ordering
2. [AV-222515]: AKO-Gateway: Gateway status conditions are not reflected correctly on deleting certificate associated with gateway

TODO: It doesn't take care of deleting Parent VS if no certificate attached to it. Parent VS will be there with default-cert.

### CODE REVIEW

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

File: ako-gateway-api/k8s/validator.go [hunk 1]
```diff
@@ -318,3 +322,119 @@ func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
...
+			}
+		} // gw loop ends
+		// update the mapping
+		akogatewayapiobjects.GatewayApiLister().UpdateSecretToGateway(secretNSName, clonedGWList)
+	}
+}
+
+// Mapping has to be taken care between secret and gateway
+func setConditions(gwStatus *gatewayv1.GatewayStatus, index int, generation int64, isDelete bool) {
+	if !isDelete {
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:384]: Can we restructure it in if{} else {} construct ?

File: ako-gateway-api/k8s/validator.go [hunk 2]
```diff
@@ -318,3 +322,119 @@ func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
...
+				clonedGWList = utils.Remove(clonedGWList, gwNSName)
+			}
+		} // gw loop ends
+		// update the mapping
+		akogatewayapiobjects.GatewayApiLister().UpdateSecretToGateway(secretNSName, clonedGWList)
+	}
+}
+
+// Mapping has to be taken care between secret and gateway
+func setConditions(gwStatus *gatewayv1.GatewayStatus, index int, generation int64, isDelete bool) {
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:383]: Should this function be renamed to setListenerConditions ?

File: ako-gateway-api/k8s/validator.go [hunk 3]
```diff
@@ -318,3 +322,119 @@ func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
...
+	akogatewayapistatus.NewCondition().
+		Type(string(gatewayv1.ListenerConditionResolvedRefs)).
+		Reason(string(gatewayv1.ListenerReasonInvalidCertificateRef)).
+		Message("Secret does not exist").
+		Status(metav1.ConditionFalse).
+		ObservedGeneration(generation).SetIn(&gwStatus.Listeners[index].Conditions)
+}
+
+func setGatewayCondition(gwStatus *gatewayv1.GatewayStatus, validListenerCount, totalListenerCount int, observedGeneration int64) {
+	if validListenerCount == 0 {
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:416]: can be restructured as if{.....}elif{.....}else{....}

File: tests/gatewayapitests/utils.go [hunk 4]
```diff
@@ -176,7 +176,52 @@ func GetListenersOnHostname(hostnames []string) []gatewayv1.Listener {
...
+				Type:               string(gatewayv1.GatewayConditionAccepted),
+				Status:             metav1.ConditionFalse,
+				Message:            "Gateway does not contain any valid listener",
+				ObservedGeneration: 1,
+				Reason:             string(gatewayv1.GatewayReasonListenersNotValid),
+			},
+		},
+		Listeners: GetListenerStatusV1(ports, []int32{0, 0}, true, false),
+	}
+	expectedStatus.Listeners[0].Conditions[0].Reason = string(gatewayv1.ListenerReasonInvalid)
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply tests/gatewayapitests/utils.go:192]: it is already set in `GetListenerStatusV1`

File: tests/gatewayapitests/utils.go [hunk 5]
```diff
@@ -176,7 +176,52 @@ func GetListenersOnHostname(hostnames []string) []gatewayv1.Listener {
...
+				Type:               string(gatewayv1.GatewayConditionAccepted),
+				Status:             metav1.ConditionTrue,
+				Message:            "Gateway configuration is valid",
+				ObservedGeneration: 1,
+				Reason:             string(gatewayv1.GatewayConditionAccepted),
+			},
+		},
+		Listeners: GetListenerStatusV1(ports, []int32{0, 0}, true, false),
+	}
+	expectedStatus.Listeners[0].Conditions[0].Reason = string(gatewayv1.ListenerReasonAccepted)
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply tests/gatewayapitests/utils.go:216]: already set in `GetListenerStatusV1`

File: tests/gatewayapitests/graphlayer/gateway_test.go [hunk 6]
```diff
@@ -428,10 +428,8 @@ func TestSecretCreateDelete(t *testing.T) {
 		found, _ := objects.SharedAviGraphLister().Get(modelName)
 		return found
 	}, 30*time.Second).Should(gomega.Equal(false))
-
-	_, aviModel := objects.SharedAviGraphLister().Get(modelName)
-	g.Expect(aviModel).To(gomega.BeNil())
-
+	// add delay
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply tests/gatewayapitests/graphlayer/gateway_test.go:432]: already been checked in 428-430

File: ako-gateway-api/k8s/validator.go [hunk 7]
```diff
@@ -318,3 +322,119 @@ func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
...
+			Status(metav1.ConditionFalse).
+			ObservedGeneration(observedGeneration).SetIn(&gwStatus.Conditions)
+		return
+	}
+	if validListenerCount < totalListenerCount {
+		akogatewayapistatus.NewCondition().
+			Type(string(gatewayv1.GatewayConditionAccepted)).
+			Reason(string(gatewayv1.GatewayReasonListenersNotValid)).
+			Message("Gateway contains atleast one valid listener").
+			Status(metav1.ConditionFalse).
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go]: done

File: ako-gateway-api/k8s/validator.go (hunk 3 above)
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go:416]: this should be fine

File: ako-gateway-api/k8s/validator.go (hunk 1 above)
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go:384]: this should be fine.

File: ako-gateway-api/k8s/validator.go (hunk 2 above)
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go]: done

**Comment by akshayhavile** [comment]
build ako

**Review by pkoshtavmware — APPROVED** [review]
LGTM

Please attach Ft run results

**Comment by akshayhavile** [comment]
```+----------------------------------------------------------------------------------+----------------+---------------------+-------------+-------------+
| Absolute Path:Test case                                                          |     Results    |       Duration      |  Start Time |   End Time  |
+----------------------------------------------------------------------------------+----------------+---------------------+-------------+-------------+
|      Test Suite Path and Name : avitest/functional/ako/ako-gw-                   |                |                     |             |             |
| api/test_gateway.py                                                              |                |                     |             |             |
| ++++++++++++++++++++++++++++++++++++++++                                         |                |                     |             |             |
| TestGateway.test_setup_configuration                                             |   [1;32m  PASSED  [0m   |       00:00:04      |   03:46:20  |   03:46:24  |
| TestGateway.test_gateway_crud                                                    |   [1;32m  PASSED  [0m   |       00:02:47      |   03:46:24  |   03:49:12  |
| TestGateway.test_update_gateway_without_gateway_class                            |   [1;32m  PASSED  [0m   |       00:02:33      |   03:49:12  |   03:51:45  |
| TestGateway.test_gateway_with_different_hosts                                    |   [1;32m  PASSED  [0m   |       00:02:17      |   03:51:45  |   03:54:02  |
| TestGateway.test_parent_vs_from_secure_gateway                                   |   [1;32m  PASSED  [0m   |       00:02:17      |   03:54:02  |   03:56:19  |
| TestGateway.test_parent_vs_from_insecure_gateway                                 |   [1;32m  PASSED  [0m   |       00:02:02      |   03:56:19  |   03:58:22  |
| TestGateway.test_gateway_with_static_ip_address                                  |   [1;32m  PASSED  [0m   |       00:01:17      |   03:58:22  |   03:59:39  |
| TestGateway.test_gateway_with_invalid_secret_ref                                 |   [1;32m  PASSED  [0m   |       00:02:47      |   03:59:39  |   04:02:26  |
| TestGateway.test_delete_configurations                                           |   [1;32m  PASSED  [0m   |       00:00:11      |   04:02:26  |   04:02:37  |
| -------------------------------------------------------------------------------- |  ------------  |     -----------     | ----------- | ----------- |
|                                                                                  |  [1;32m  Passed: 9 [0m  |                     |             |             |
|                                                                                  |    [1;31mFailed:0[0m    |                     |             |             |
|  Total test cases: 9                                                             |    [1;33mSkipped:0[0m   | Total Time:00:16:17 |             |             |
+----------------------------------------------------------------------------------+----------------+---------------------+-------------+-------------+```

File: ako-gateway-api/k8s/validator.go [hunk 8]
```diff
@@ -318,3 +322,119 @@ func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
 	}
 	return true
 }
+
+func ValidateGatewayListenerWithSecret(namespace, name string, deleteFlag bool) {
+	secretNSName := namespace + "/" + name
+	present, gwList := akogatewayapiobjects.GatewayApiLister().GetSecretToGateway(secretNSName)
+
+	if present {
```
Comment (ashishnayak96) [inline ako-gateway-api/k8s/validator.go:330]: nit: if !present { return } will decrease code complexity

File: ako-gateway-api/lib/lib.go [hunk 9]
```diff
@@ -151,7 +151,12 @@ func FindTargetPort(serviceName, ns string, svcPort int32, key string) intstr.In
 	}
 	return intstr.IntOrString{}
 }
-
+func IsListenerInvalid(gwStatus *gatewayv1.GatewayStatus, listenerIndex int) bool {
+	if len(gwStatus.Listeners) > int(listenerIndex) && gwStatus.Listeners[listenerIndex].Conditions[0].Type == string(gatewayv1.ListenerConditionAccepted) && gwStatus.Listeners[listenerIndex].Conditions[0].Status == "False" {
```
Comment (ashishnayak96) [inline ako-gateway-api/lib/lib.go:155]: listenerIndex is already of type int, we dont need to type convert it to int again.
can conditions[0] create index out of range error?

File: ako-gateway-api/nodes/dequeue_ingestion.go [hunk 10]
```diff
@@ -70,7 +70,17 @@ func DequeueIngestion(key string, fullsync bool) {
...
 
 		modelFound, modelIntf := objects.SharedAviGraphLister().Get(modelName)
-		if !modelFound || modelIntf == nil {
+		// Seq: GW first and the secret created.
+		modelNil := !modelFound || modelIntf == nil
+		if objType == utils.Secret && modelNil {
+			handleGateway(parentNs, parentName, fullsync, key)
+			modelFound, modelIntf = objects.SharedAviGraphLister().Get(modelName)
+			modelNil = !modelFound || modelIntf == nil
+			if modelNil {
```
Comment (ashishnayak96) [inline ako-gateway-api/nodes/dequeue_ingestion.go:79]: this if condition will always evaluate true

File: ako-gateway-api/nodes/dequeue_ingestion.go [hunk 11]
```diff
@@ -70,7 +70,17 @@ func DequeueIngestion(key string, fullsync bool) {
 		modelName := lib.GetModelName(lib.GetTenant(), akogatewayapilib.GetGatewayParentName(parentNs, parentName))
 
 		modelFound, modelIntf := objects.SharedAviGraphLister().Get(modelName)
-		if !modelFound || modelIntf == nil {
+		// Seq: GW first and the secret created.
+		modelNil := !modelFound || modelIntf == nil
+		if objType == utils.Secret && modelNil {
```
Comment (ashishnayak96) [inline ako-gateway-api/nodes/dequeue_ingestion.go:75]: is modelNil check intended?

**Review by ashishnayak96 — APPROVED** [review]
LGTM, minor comments

File: ako-gateway-api/lib/lib.go (hunk 9 above)
Comment (akshayhavile) [reply ako-gateway-api/lib/lib.go:155]: Will take care of your comment in next.

**Pull request merged**